src/transformers/models/big_bird/modeling_big_bird.py

```diff
@@ -1055,6 +1116,29 @@ def _get_rand_attn_plan(from_seq_length, from_block_size, num_rand_blocks):
 
         return plan_from_length, plan_num_rand_blocks
 
+    def _bigbird_block_rand_mask_fast(self, from_seq_length, from_block_size, num_rand_blocks, n_heads):
+        """
+        A fast path to create adjacency list of random attention. The permutation list is pre-computed by
+        self.generate_rand_attn_tables and stored in self.rand_attn_tables. This function choose one of the adjacency
+        list from the pre-computed list.
+
+        Args:
+            from_seq_length: int. length of from sequence.
+            from_block_size: int. size of block in from sequence.
+            num_rand_blocks: int. Number of random chunks per row.
+
+        Returns:
+            adjacency list of size from_seq_length//from_block_size-2 by num_rand_blocks
+        """
+        rand_attn = [
+            np.zeros((from_seq_length // from_block_size - 2, num_rand_blocks), dtype=np.int32) for _ in range(n_heads)
+        ]
```

Comment:
we can just do 
```python 
>>> # sample n_heads random indices, for from_seq_length // from_block_size nb of blocks
>>> rand_i = np.random.randint(len(self.rand_attn_tables[i - 1]), size=(from_seq_length // from_block_size - 2, num_rand_blocks))
>>> # get the masks by broadcasting
>>> self.rand_attn_tables[rand_i]
```
and get rid of the function entirely 😄 
Also, the shape of the returned `adjacency list` is `nheads X (from_seq_length // from_block_size - 2 X num_rand_blocks)` which is not said in the documentation. 
This can only work if you have numpy arrays or torch tensors. It think it makes sens to use tensors.

Reply:
Oh, I see what your points are. will fix in the next commit.

Reply:
Hi @ArthurZucker , I just rethought our previous discussion and found the function you proposed can't replace the original function.  The pre-generated `self.rand_attn_tables`  has `from_seq_length // from_block_size - 2` items. Each item has its own permutations. For example, item 0 or self.rand_attn_tables[0]'s all permutations generated by the rule that  1) select num_rand_blocks items from a range such as middle_seq[2, last], and  2) generate all possible orders of the combinations of these num_rand_blocks items. So, the length of each item in self.rand_attn_tables may be different. 

Since these permutations are pre-generated, we only need to randomly choose one of the permutations for all items in self.rand_attn_tables. For example, for item 0 we need to generate one integer whose range is [0, len(self.rand_attn_tables[0])]. It is an index of self.rand_attn_tables[0] and it represents which permutation we will choose for item 0. We need to generate `(from_seq_length // from_block_size - 2)` random indices. In totally, it has `(from_seq_length // from_block_size - 2) * num_rand_blocks` integers after we use generated indices to find the corresponding permutations. 

 I also remove the original nhead calls to `_bigbird_block_rand_mask`. So the final returned rand_attn has nhead items. Each item includes `(from_seq_length // from_block_size - 2) * num_rand_blocks` numbers. 

Because the length of each item in self.rand_attn_tables may be different, the first argument of np.random.randint() can't be a fixed value. We still need a for loop to go through all item [0, from_seq_length // from_block_size - 2]. We can't use `size=(from_seq_length // from_block_size - 2, num_rand_blocks)` as the second argument because 1)  the length of each item in self.rand_attn_tables may be different, 2) we already generate num_rand_blocks integers in our pre-generated self.rand_attn_tables.

Reply:
Okay, the only suggestion now is the following: 
```python 
def _bigbird_block_rand_mask_fast(self, from_seq_length, from_block_size, num_rand_blocks, n_heads):
    rand_attn = np.zeros((n_heads, from_seq_length // from_block_size - 2, num_rand_blocks), dtype=np.int32)

    for i in range(1, from_seq_length // from_block_size - 1):
        rand_i = np.random.randint(len(self.rand_attn_tables[i - 1]), size=n_heads)
        rand_attn[:, i - 1] = self.rand_attn_tables[i - 1][rand_i]

    return rand_attn
```
The second for loop is useless for updating a slice.

Reply:
Got an error in the following:
```
  File "/home/yhao24/anaconda3/envs/pt_may30/lib/python3.10/site-packages/transformers/models/big_bird/modeling_big_bird.py", line 1145, in _bigbird_block_rand_mask_fast
    rand_attn[:, i - 1] = self.rand_attn_tables[i - 1][rand_i]
TypeError: only integer scalar arrays can be converted to a scalar index
```
Because the `rand_i` is a list, we can't use it as an index.